clean-modular-code/python-pep-8.md

```diff
@@ -224,27 +228,103 @@ plt.show()
+::::
+
+The pyOpenSci community has been slowly adopting Ruff because it can run many code formatters with a single configuration file, making it a convenient and easy-to-use tool. 
+
+:::{admonition} Linter vs. code formatter
+:class: tip
+
+A code formatter is a tool that helps you modify your code to follow code standards. A linter checks your code and tells you if things need to be fixed, but it generally won't fix them for you.  
+
+Ruff is both a linter and code formatter which is why we like to use it when developing both software and .py file based workflows. 
+:::
 
 +++ {"editable": true, "slideshow": {"slide_type": ""}}
 
-## Additional Resources
+DRAFT!! 
+
+
+## Best Practices for Importing Python Packages In Scientific Code
+
+There are a set of best practices that you should follow when importing **Python** packages in your code. These best practices are outlined in the <a href="https://www.python.org/dev/peps/pep-0008/#imports" target="_blank">PEP 8 guidelines</a> and apply to both **Python** scripts and to working in **Jupyter Notebook** files.
+
+## Import Python Libraries at the Top of Your Script or Notebook
+
+It is good practice to import all of the packages that you will need at the top of your **Python** script (.py file) or in the first code cell of a **Jupyter Notebook** file. 
+
+This allows anyone looking at your code to immediately know what packages they need to have installed in order to successfully run the code. This rule also follows the PEP 8 conventions for **Python** code.
+
+```python
+import os 
+
+import pandas as pd
+import numpy as np
+```
+
+Once you have imported all of the packages that you need to run your code in a script, you have access to all of the functions and classes defined in each package. 
+
+If these imports are at the top of the script or **Jupyter Notebook** file, then you will be able to use those packages in any code lines that follow. 
+
+This means that if you import a package *after* running some code that requires that package, your code will not run successfully.
+
+
+## 2. List Package Imports Following PEP 8 Standards: Most Common First, Followed By Third Party
+
+<a href="https://www.python.org/dev/peps/pep-0008/#imports" target="_blank">PEP 8</a> also specifies the order in which you should list your imports as follows (starting with the most commonly used):
+
+> Imports should be grouped in the following order:
+>    Standard library imports.
+>    Related third party imports.
+>    Local application/library specific imports.
+> You should put a blank line between each group of imports.
+
+You may be wondering, what is a standard library import? The standard library imports are commonly used tools that are general in purpose and are part of the standard library of **Python**. These including things like:
```

Comment:
"that are included with python," "that you don't have to install," with link to https://docs.python.org/3/library/index.html ?